array.c

```diff
@@ -212,6 +212,33 @@ ary_embeddable_p(long capa)
 #endif
 }
 
+bool
+rb_ary_pool_moveable_p(VALUE ary)
+{
+    // if the array is shared or a shared root then it's not moveable
+    return !(ARY_SHARED_P(ary) || ARY_SHARED_ROOT_P(ary));
+}
+
+void ary_verify(VALUE ary);
+
+size_t
+rb_ary_size_as_embedded(VALUE ary)
+{
+    size_t real_size;
+#if USE_RVARGC
+    if (ARY_EMBED_P(ary)) {
+        real_size = ary_embed_size(ARY_EMBED_LEN(ary));
+    } else if (rb_ary_pool_moveable_p(ary)) {
+        real_size = ary_embed_size(ARY_HEAP_CAPA(ary));
+    } else {
+#endif
+        real_size = sizeof(struct RString);
+#if USE_RVARGC
+    }
+#endif
```

Comment:
Do we need separate logic for whether or not `USE_RVARGC` is turned on? When `USE_RVARGC` is false `gc_compact_destination_pool` will automatically keep the object in the first size pool.